orchestrator/src/coordinator.rs

```diff
@@ -1,30 +1,128 @@
+    id: AtomicU64,
+}
+
+impl Container {
+    async fn fanout(mut command_rx: mpsc::Receiver<FanoutCommand>, mut rx: mpsc::Receiver<ContainerMessage>) {
+        let mut waiting = HashMap::new();
+
+        loop {
+            select! {
+                command = command_rx.recv() => {
+                    let command = command.expect("Handle this");
+                    match command {
+                        FanoutCommand::Listen(id, waiter) => {
+                            waiting.insert(id, waiter);
+                            // TODO: ensure not replacing
+                        }
+                    }
+                },
+
+                msg = rx.recv() => {
+                    let msg = msg.expect("Handle this");
+                    let id = 0; // TODO: some uniform way of getting the ID for _any_ message
+                    if let Some(waiter) = waiting.get(&id) {
+                        waiter.send(msg).await.ok(/* Don't care about it */);
+                    }
+                    // TODO: log unattended messages?
+                }
+            }
+        }
+    }
+
+    pub async fn compile(&self, request: CompileRequest) -> Result<CompileResponse> {
+        let (mut result, mut stdout, mut stderr) = self.begin_compile(request).await?;
+
+        let mut stdout_buf = String::new();
+        let mut stderr_buf = String::new();
+
+        loop {
+            select! {
+                result = &mut result => {
+                    let mut result = result.expect("handle me")?;
+                    result.stdout = stdout_buf;
+                    result.stderr = stderr_buf;
+                    return Ok(result);
+                },
+
+                Some(stdout) = stdout.recv() => stdout_buf.push_str(&stdout),
+                Some(stderr) = stderr.recv() => stderr_buf.push_str(&stderr),
+            }
+        }
+    }
+
+    pub async fn begin_compile(&self, request: CompileRequest) -> Result<(
+        JoinHandle<Result<CompileResponse>>, // TODO: shouldn't include stdout / stderr
+        mpsc::Receiver<String>,
+        mpsc::Receiver<String>,
+    )> {
+        let Self { child, tx: to_worker_tx, command_tx, id } = self;
+        let id = id.fetch_add(1, Ordering::SeqCst);
+
+        let (from_worker_tx, mut from_worker_rx) = mpsc::channel(8);
+
+        command_tx.send(FanoutCommand::Listen(id, from_worker_tx)).await.unwrap();
+        to_worker_tx.send(PlaygroundMessage::Request(id, HighLevelRequest::Compile(request))).await.expect("handle this");
+
+        let (stdout_tx, stdout_rx) = mpsc::channel(8);
+        let (stderr_tx, stderr_rx) = mpsc::channel(8);
+
+        let x = tokio::spawn(async move {
+            while let Some(container_msg) = from_worker_rx.recv().await {
+                match container_msg {
+                    ContainerMessage::Response(_, resp) => {
+                        match resp {
+                            HighLevelResponse::Compile(resp) => return Ok(resp),
```

Comment:
I'd actually expect that we might not send `CompileRequest` / `CompileResponse` here; what about this function being responsible for lowering / lifting?

We'd get the `CompileRequest` here, break that down into something like:

1. `WriteFile` 
2. `StartProcess`
3. (Stream stdin/out/err and wait for ProcessComplete)
4. `ReadFile`

It's possible that the only messages between the coordinator and the worker are those about reading/writing files and process-related (start, kill, stopped, stdin/out/err).